Move `const` `Array1D`/`2D` to `constexpr std::array`

mitchute opened the pull request:
Pull request overview
---------------------
 - Moves `const Array1D` and `Array2D` to `constexpr std::array`
 - To be merged after #9166 
 - Fixes #9039

mitchute committed:
moving const Array1D to contexper std::array

mitchute committed:
moving const Array1D to contexpr std::array

amirroth commented on src/EnergyPlus/DataPhotovoltaics.hh:
```diff
@@ -60,6 +60,8 @@ namespace EnergyPlus {
+
     enum class PVModel
     {
         Unassigned,
```
`Unassigned` needs to be explicitly assigned to -1, otherwise it will be assigned to 0 by default.

mitchute committed:
moving const Array1D to contexpr std::array

Myoldmopar commented:
Is there more work to come on this draft PR @mitchute ?  Looks like another great round of changes!

mitchute commented:
Yeah, I've got another 80 or so on my list to chew through still. Probably won't be done this week. I'm deliberately taking it slow here anyway since this has the potential to break things, so I'd like to have pretty consistent CI testing between rounds of changes.

amirroth commented on src/EnergyPlus/SurfaceGeometry.cc:
```diff
@@ -112,6 +112,10 @@ namespace SurfaceGeometry {
 
+    int constexpr UnenteredAdjacentZoneSurface = -998; // allows users to enter one zone surface ("Zone")
+                                                       // referencing another in adjacent zone
+    int constexpr UnreconciledZoneSurface = -999;      // interim value between entering surfaces ("Surface") and reconciling
```
Put a pin in this thing. The only negative `enum` like value we should be using is -1 to indicated `Invalid`, `Unassigned`, `Error`, etc. Using random negative integers to mean different things is bad design.

mitchute committed:
Merge branch 'develop' into constexpr-const-arrays

mitchute committed:
moving const Array1D to contexpr std::array

mitchute committed:
moving const Array1D to contexpr std::array

mitchute committed:
moving const Array1D to contexpr std::array

mitchute committed:
add  to all

mitchute committed:
collapse several constexpr arrays into single array, collapse logic

mitchute committed:
moving const Array1D to contexpr std::array, refactoring some of DataSurfaceColors and adding a unit test

mitchute committed:
move final 'const Array1Ds' to 'constexpr std::array', minor refactoring in HVACManager to eliminate dependence on ObjexxFCL 'sum'

mitchute committed:
bonus points - convert const Array2D's to constexpr std::arrays

mitchute commented on src/EnergyPlus/HVACManager.cc:
```diff
@@ -1105,47 +1107,46 @@ void SimHVAC(EnergyPlusData &state)
 
                     for (NodeIndex = 1; NodeIndex <= state.dataConvergeParams->ZoneInletConvergence(ZoneNum).NumInletNodes; ++NodeIndex) {
 
+                        auto &humRatInletNode = state.dataConvergeParams->ZoneInletConvergence(ZoneNum).InletNode(NodeIndex).HumidityRatio;
+                        auto &mdotInletNode = state.dataConvergeParams->ZoneInletConvergence(ZoneNum).InletNode(NodeIndex).MassFlowRate;
+                        auto &inletTemp = state.dataConvergeParams->ZoneInletConvergence(ZoneNum).InletNode(NodeIndex).Temperature;
+
```
Adding local references to these variables in a few places to clean up how the code looks. I'm not stuck on the names, so let me know if you'd like to see something else.

mitchute commented on src/EnergyPlus/HVACManager.cc:
```diff
@@ -1105,47 +1107,46 @@ void SimHVAC(EnergyPlusData &state)
+                                    std::begin(ConvergLogStackARR), std::end(ConvergLogStackARR), std::begin(humRatInletNode), 0.0);
+
                                 SlopeHumRat =
-                                    (sum_ConvergLogStackARR *
-                                         sum(state.dataConvergeParams->ZoneInletConvergence(ZoneNum).InletNode(NodeIndex).HumidityRatio) -
-                                     double(DataConvergParams::ConvergLogStackDepth) *
-                                         sum((ConvergLogStackARR *
-                                              state.dataConvergeParams->ZoneInletConvergence(ZoneNum).InletNode(NodeIndex).HumidityRatio))) /
+                                    (sum_ConvergLogStackARR * sum(humRatInletNode) -
+                                     double(DataConvergParams::ConvergLogStackDepth) * humRatInletNodDotProd) /
```
It was hard to see, but the old code was using Objexx `sum` here, but this part is just a dot product, so I pulled that out into a standalone variable here to make this easier to read. (Plus Objexx `sum` won't work with an `Array1D` and `std::array`, so I had to use an STL function any way)

mitchute commented on src/EnergyPlus/HVACManager.cc:
```diff
@@ -1204,42 +1200,37 @@ void SimHVAC(EnergyPlusData &state)
                             }
                             if (!FoundOscillationByDuplicate) {
+
+                                auto humRatInletNodDotProd =
+                                    std::inner_product(std::begin(ConvergLogStackARR), std::end(ConvergLogStackARR), std::begin(mdotInletNode), 0.0);
```
Another dot product pulled out here.

mitchute commented on src/EnergyPlus/HVACManager.cc:
```diff
@@ -1296,42 +1283,36 @@ void SimHVAC(EnergyPlusData &state)
                             }
                             if (!FoundOscillationByDuplicate) {
+
+                                auto inletTempDotProd =
+                                    std::inner_product(std::begin(ConvergLogStackARR), std::end(ConvergLogStackARR), std::begin(inletTemp), 0.0);
```
Another here.

mitchute commented on src/EnergyPlus/HVACManager.cc:
```diff
@@ -665,10 +666,11 @@ void SimHVAC(EnergyPlusData &state)
-    Real64 const sum_ConvergLogStackARR(sum(ConvergLogStackARR));
-    Real64 const square_sum_ConvergLogStackARR(pow_2(sum_ConvergLogStackARR));
-    Real64 const sum_square_ConvergLogStackARR(sum(pow(ConvergLogStackARR, 2)));
+    static constexpr std::array<Real64, DataConvergParams::ConvergLogStackDepth> ConvergLogStackARR = {
```
Most of these changes are self-explanatory, but there are a few that could use a look-over by reviewers. These HVACManager changes are one of those places.

mitchute commented on src/EnergyPlus/HeatBalanceManager.cc:
```diff
@@ -2510,12 +2510,12 @@ namespace HeatBalanceManager {
             if (GasType >= 1 && GasType <= 4) {
-                state.dataMaterial->Material(MaterNum).GasWght(1) = GasWght(GasType);
-                state.dataMaterial->Material(MaterNum).GasSpecHeatRatio(1) = GasSpecificHeatRatio(GasType);
+                state.dataMaterial->Material(MaterNum).GasWght(1) = GasWght[GasType - 1];
+                state.dataMaterial->Material(MaterNum).GasSpecHeatRatio(1) = GasSpecificHeatRatio[GasType - 1];
```
I tried to minimize adding `- 1` to all of these index lookups where possible. In some places, though, it just wasn't feasible to do the required changes to make that go away.

mitchute commented on src/EnergyPlus/StandardRatings.cc:
```diff
@@ -125,64 +125,54 @@ namespace StandardRatings {
-    Array1D<Real64> const RegionSixFracBinHoursAtOutdoorBinTemp(
-        18, {0.113, 0.206, 0.215, 0.204, 0.141, 0.076, 0.034, 0.008, 0.003, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0});
+    static constexpr std::array<Real64, 18> OutdoorBinTemperature = {
+        16.67, 13.89, 11.11, 8.33, 5.56, 2.78, 0.00, -2.78, -5.56, -8.33, -11.11, -13.89, -16.67, -19.44, -22.22, -25.00, -27.78, -30.56};
+
+    // Fractional bin hours for different bin temperatures from ANSI/AHRI 210/240
+    static constexpr std::array<std::array<Real64, 18>, 6> FracBinHoursAtOutdoorBinTemp = {
+        {{0.291, 0.239, 0.194, 0.129, 0.081, 0.041, 0.019, 0.005, 0.001, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0},
+         {0.215, 0.189, 0.163, 0.143, 0.112, 0.088, 0.056, 0.024, 0.008, 0.002, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0},
+         {0.153, 0.142, 0.138, 0.137, 0.135, 0.118, 0.092, 0.047, 0.021, 0.009, 0.005, 0.002, 0.001, 0.0, 0.0, 0.0, 0.0, 0.0},
+         {0.132, 0.111, 0.103, 0.093, 0.1, 0.109, 0.126, 0.087, 0.055, 0.036, 0.026, 0.013, 0.006, 0.002, 0.001, 0.0, 0.0, 0.0},
+         {0.106, 0.092, 0.086, 0.076, 0.078, 0.087, 0.102, 0.094, 0.074, 0.055, 0.047, 0.038, 0.029, 0.018, 0.01, 0.005, 0.002, 0.001},
+         {0.113, 0.206, 0.215, 0.204, 0.141, 0.076, 0.034, 0.008, 0.003, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0}}};
```
These size `BinHoursAtHoursAtOutdoorBinTemp` arrays were all collapsed into a single array. You'll see why later.

mitchute commented on src/EnergyPlus/StandardRatings.cc:
```diff
@@ -1410,55 +1382,43 @@ namespace StandardRatings {
 
-        for (BinNum = 1; BinNum <= TotalNumOfTemperatureBins(RegionNum); ++BinNum) {
-
-            if (RegionNum == 1) {
-                FractionalBinHours = RegionOneFracBinHoursAtOutdoorBinTemp(BinNum);
-            } else if (RegionNum == 2) {
-                FractionalBinHours = RegionTwoFracBinHoursAtOutdoorBinTemp(BinNum);
-            } else if (RegionNum == 3) {
-                FractionalBinHours = RegionThreeFracBinHoursAtOutdoorBinTemp(BinNum);
-            } else if (RegionNum == 4) {
-                FractionalBinHours = RegionFourFracBinHoursAtOutdoorBinTemp(BinNum);
-            } else if (RegionNum == 5) {
-                FractionalBinHours = RegionFiveFracBinHoursAtOutdoorBinTemp(BinNum);
-            } else if (RegionNum == 6) {
-                FractionalBinHours = RegionSixFracBinHoursAtOutdoorBinTemp(BinNum);
-            }
```
Collapsing all of those arrays into a single array allows us to eliminate these IF/ELSE ladders for lookups. 🎩  tip @amirroth for spotting that.

mitchute commented on src/EnergyPlus/StandardRatings.cc:
```diff
@@ -1410,55 +1382,43 @@ namespace StandardRatings {
 
-            BuildingLoad =
-                (18.33 - OutdoorBinTemperature(BinNum)) / (18.33 - OutdoorDesignTemperature(RegionNum)) * CorrectionFactor * DesignHeatingRequirement;
+            FractionalBinHours = FracBinHoursAtOutdoorBinTemp[RegionNum - 1][BinNum];
```
Here's the new lookup.

mitchute commented on src/EnergyPlus/ThermalComfort.cc:
```diff
@@ -1882,12 +1882,12 @@ namespace ThermalComfort {
         X += H2;
 
-        for (J = 1; J <= 2; ++J) {
+        for (J = 0; J < 2; ++J) {
```
I changed the bounds on loops such as this where possible to eliminate the 1-based vs 0-based `- 1` offset.

mitchute commented on src/EnergyPlus/ThermalComfort.cc:
```diff
@@ -2957,7 +2957,7 @@ namespace ThermalComfort {
         // SUBROUTINE PARAMETER DEFINITIONS:
         static Real64 constexpr alpha(0.8);
-        static Array1D<Real64> const alpha_pow({pow_6(alpha), pow_5(alpha), pow_4(alpha), pow_3(alpha), pow_2(alpha), alpha, 1.0}); // alpha^(7-i)
+        static constexpr std::array<Real64, 7> alpha_pow = {0.262144, 0.32768, 0.4096, 0.512, 0.64, 0.8, 1.0}; // alpha^(6-0)
```
I just hardcoded these values here. If `alpha` ever changes, this will also have to change. There are a few places like that throughout the code. The array history stuff at the top of HVACManager comes to mind, but there may be others.

mitchute commented on src/EnergyPlus/WindowComplexManager.cc:
```diff
@@ -2838,13 +2838,13 @@ namespace WindowComplexManager {
+        GasType = static_cast<int>(DataComplexFenestration::GasCoeffs::Air);
+        wght(iprop(1, 1)) = GasWght[GasType - 1];
+        gama(iprop(1, 1)) = GasSpecificHeatRatio[GasType - 1];
         for (ICoeff = 1; ICoeff <= 3; ++ICoeff) {
-            gcon(ICoeff, iprop(1, 1)) = GasCoeffsCon(ICoeff, static_cast<int>(GasType));
-            gvis(ICoeff, iprop(1, 1)) = GasCoeffsVis(ICoeff, static_cast<int>(GasType));
-            gcp(ICoeff, iprop(1, 1)) = GasCoeffsCp(ICoeff, static_cast<int>(GasType));
+            gcon(ICoeff, iprop(1, 1)) = GasCoeffsCon[ICoeff - 1][GasType - 1];
+            gvis(ICoeff, iprop(1, 1)) = GasCoeffsVis[ICoeff - 1][GasType - 1];
+            gcp(ICoeff, iprop(1, 1)) = GasCoeffsCp[ICoeff - 1][GasType - 1];
```
I just made `GasType` and int here and went with one `static_cast` instead of five.

mitchute commented on tst/EnergyPlus/unit/CMakeLists.txt:
```diff
@@ -17,16 +17,16 @@ set(test_src
     Fixtures/SQLiteFixture.hh
     PhaseChangeModeling/HysteresisModel.unit.cc
     AdvancedAFN.unit.cc
-    AirflowNetworkSimulationControl.unit.cc
```
I added one new file here, but the rest of the changes are just due to sorting the lines.

mitchute commented on tst/EnergyPlus/unit/DataSurfaceColors.unit.cc:
```diff
@@ -45,21 +45,36 @@
 
-// ObjexxFCL Headers
-#include <ObjexxFCL/Array.functions.hh>
+// Google Test Headers
```
New file and unit test here.

mitchute commented on src/EnergyPlus/DataSurfaceColors.cc:
```diff
@@ -79,15 +79,34 @@ bool MatchAndSetColorTextString(EnergyPlusData &state,
-        }
-    }
-    return WasSet;
+    static constexpr std::array<std::string_view, static_cast<int>(DataSurfaceColors::ColorNo::NUM)> colorkeys = {
+        "TEXT",
+        "WALLS",
+        "WINDOWS",
+        "GLASSDOORS",
+        "DOORS",
+        "ROOFS",
+        "FLOORS",
+        "DETACHEDBUILDINGSHADES",
+        "DETACHEDFIXEDSHADES",
+        "ATTACHEDBUILDINGSHADES",
+        "PHOTOVOLTAICS",
+        "TUBULARDAYLIGHTDOMES",
+        "TUBULARDAYLIGHTDIFFUSERS",
+        "DAYLIGHTREFERENCEPOINT1",
+        "DAYLIGHTREFERENCEPOINT2",
+    };
+
+    // ColorType must be "DXF"
+    if (ColorType != "DXF") return false;
+
+    // try to find enum value
+    int foundIdx = getEnumerationValue(colorkeys, UtilityRoutines::MakeUPPERCase(String));
+    if (foundIdx == -1) return false;
+
+    // if we've made it here, we found the value
+    state.dataSurfColor->DXFcolorno[foundIdx] = SetValue;
+    return true;
```
I refactored this function because I'm pretty sure it was broken before. `FindItem` uses a 1-based lookup, but these Array1Ds were definitely 0-based before. This and the new unit test could use an extra 5 seconds from reviewers.

mitchute reviewed (commented):
Ready for review here

mitchute committed:
remove some unused Objecxx functions, fix a few constexpr that were missed

mitchute commented:
OK, I removed `reshape2` and a few other unused Objexx functions. There were also a few other constexpr's that were missed, so those were also cleaned up in the latest commit.

mitchute committed:
add custom check script, fix another const array

mitchute committed:
Merge branch 'develop' into constexpr-const-arrays

mitchute committed:
add license

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into constexpr-const-arrays

Myoldmopar committed:
Refactor constexpr check into python unittest

Myoldmopar commented on scripts/dev/check_constexpr.py:
```diff
@@ -0,0 +1,431 @@
+    errors_found &= constexpr_check(tst_path)
+
+    if errors_found:
+        raise sys.exit(1)
```
This is a really great addition.  I modified it locally to take advantage of Python unittest stuff, but did not change the functionality of it.  I'll push those changes with any other stuff I change or merges from develop.

mitchute replied:
I had it all using `unittest` originally but my system wouldn't let me run the tests and the actual checks together. 🤷  Fix'r up!

Myoldmopar commented on scripts/dev/custom_check.sh:
```diff
@@ -17,5 +17,6 @@ python3 "$REPO_ROOT"/scripts/dev/verify_cmake_dirs.py || EXIT_STATUS=$?
 python3 "$REPO_ROOT"/scripts/dev/find_included_cc_files.py || EXIT_STATUS=$?
 python3 "$REPO_ROOT"/scripts/dev/analyze_state.py || EXIT_STATUS=$?
 python3 "$REPO_ROOT"/scripts/dev/check_for_c_style_comments.py || EXIT_STATUS=$?
+python3 "$REPO_ROOT"/scripts/dev/check_constexpr.py || EXIT_STATUS=$?
```
Yes, all the custom_checks!!!

Myoldmopar commented on src/EnergyPlus/AirflowNetworkBalanceManager.cc:
```diff
@@ -7229,25 +7229,22 @@ void AirflowNetworkBalanceManagerData::calculateWindPressureCoeffs(EnergyPlusDat
+        {0.60, 0.54, 0.23, -0.25, -0.61, -0.55, -0.51, -0.55, -0.61, -0.25, 0.23, 0.54},
+        {0.60, 0.48, 0.04, -0.56, -0.56, -0.42, -0.37, -0.42, -0.56, -0.56, 0.04, 0.48},
+        {0.60, 0.44, -0.26, -0.70, -0.53, -0.32, -0.22, -0.32, -0.53, -0.70, -0.26, 0.44},
+    }};
```
w000t!!

Myoldmopar commented on src/EnergyPlus/EnergyPlus.hh:
```diff
@@ -141,8 +141,6 @@ using ObjexxFCL::Optional_int_const;
 using ObjexxFCL::Optional_string_const;
 using ObjexxFCL::Reference;
-using ObjexxFCL::Reference_bool;
-using ObjexxFCL::Reference_int;
```
This whole PR is just top-tier cleanups, thanks @mitchute !

Myoldmopar commented on src/EnergyPlus/Plant/DataPlant.hh:
```diff
@@ -301,16 +301,6 @@ namespace DataPlant {
         LoopType::Both,  LoopType::Both,  LoopType::Both,  LoopType::Both,  LoopType::Both,  LoopType::Plant, LoopType::Plant, LoopType::Both,
         LoopType::Both,  LoopType::Both,  LoopType::Both,  LoopType::Plant, LoopType::Plant, LoopType::Plant, LoopType::Both,  LoopType::Both};
 
-    extern Array1D<Real64> const ConvergenceHistoryARR;
```
:+1: :+1: :+1:

Myoldmopar commented on src/EnergyPlus/StandardRatings.hh:
```diff
@@ -93,37 +90,6 @@ namespace StandardRatings {
-    // indoor dry bulb temperatures for tests A, B, C and D and Classes I, II, III, and IV
-    extern Array1D<Real64> const IndoorDBTempClassI2IV;
-    // outdoor dry bulb temperatures for tests A, B, C and D
-    extern Array1D<Real64> const OutdoorDBTempAllClassA2D;
```
:smile: :smile: :smile:

Myoldmopar commented on third_party/ObjexxFCL/src/ObjexxFCL/numeric.hh:
```diff
@@ -27,136 +27,6 @@
 
 namespace ObjexxFCL {
 
-inline
```
Thinning this out is awesome!

Myoldmopar reviewed (approved):
Super happy with this!  This is a great set of changes that should be a no-impact all-gain PR.  I'll do a final build/test locally, and I'll do a quick push of my Python changes, but this should be good to go.  Thanks @mitchute !

Myoldmopar commented on scripts/dev/check_constexpr.py:
```diff
@@ -0,0 +1,483 @@
+array_us_const_pattern = re.compile(array_us_const_str)
+
+
+class TestMatching(unittest.TestCase):
```
@mitchute I just made this into a standard unittest.TestCase instance here...

Myoldmopar commented on scripts/dev/check_constexpr.py:
```diff
@@ -0,0 +1,483 @@
+
+if __name__ == "__main__":
+    print("**** Verifying script regexes ****")
+    unittest.main(exit=False)
```
...and then down here in the actual script main code I call the `unittest.main` function.  I have to pass `exit=False` here or it will abort the script entirely.  I thought about wrapping this in a command line argument like: `check_constexpr.py test` but it's fine to just execute it every time.

Myoldmopar reviewed (approved):
This is good to go afaik.  I am running final regressions locally but I anticipate it merging before Decent wastes time on it.

Myoldmopar commented:
Regressions are all clean, unfortunately Decent caught this just in time.  I'll merge it in but I won't delete the branch to avoid causing lots of red alerts here.  Thanks @mitchute !
